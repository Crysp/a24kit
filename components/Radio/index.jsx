import React from 'react';
import PropTypes from 'prop-types';
import {
    Wrapper,
    EmptyCircle,
    Circle,
} from './styled';


export default class Radio extends React.PureComponent {
    static propTypes = {
        color: PropTypes.oneOf(['blue', 'green']),
        large: PropTypes.bool,
        disabled: PropTypes.bool,
        isChecked: PropTypes.bool,
        isHighlighted: PropTypes.bool,
    };

    static defaultProps = {
        color: 'blue',
        large: false,
        disabled: false,
        isChecked: false,
        isHighlighted: false,
    };

    render() {
        const {
            color,
            large,
            disabled,
            isChecked,
            isHighlighted,
        } = this.props;
        return (
            <Wrapper large={large}>
                <EmptyCircle
                    color={color}
                    checked={isChecked}
                    highlighted={isHighlighted}
                    disabled={disabled}
                />
                <Circle
                    color={color}
                    checked={isChecked}
                    highlighted={isHighlighted}
                    large={large}
                    disabled={disabled}
                />
            </Wrapper>
        );
    }
}
