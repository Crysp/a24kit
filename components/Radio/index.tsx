import React from 'react';
import {
    Wrapper,
    EmptyCircle,
    Circle,
} from './styled';

export interface Props {
    color: 'blue' | 'green'
    large: boolean
    disabled: boolean
    isChecked: boolean
    isHighlighted: boolean
}

export default class Radio extends React.PureComponent<Props> {
    static defaultProps = {
        color: 'blue',
        large: false,
        disabled: false,
        isChecked: false,
        isHighlighted: false,
    };

    render() {
        const {
            color,
            large,
            disabled,
            isChecked,
            isHighlighted,
        } = this.props;
        return (
            <Wrapper large={large}>
                <EmptyCircle
                    color={color}
                    checked={isChecked}
                    highlighted={isHighlighted}
                    disabled={disabled}
                />
                <Circle
                    color={color}
                    checked={isChecked}
                    highlighted={isHighlighted}
                    large={large}
                    disabled={disabled}
                />
            </Wrapper>
        );
    }
}
