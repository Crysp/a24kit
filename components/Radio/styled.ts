import styled, { css } from 'styled-components';
import { Color } from '../config';
import React from "react";

interface WrapperProps {
    large?: boolean
}

export const Wrapper = styled.div<WrapperProps>`
    position: relative;
    display: inline-block;
    top: ${props => (props.large ? 'auto' : '2px')};
    width: ${props => (props.large ? 20 : 16)}px;
    height: ${props => (props.large ? 20 : 16)}px;
`;

const getEmptyCircleStyles = (props) => {
    let backgroundColor = Color.White;
    let borderColor = Color.Pale;
    let animationState = css`
        opacity: 1;
        transform: scale(1);
    `;
    if (props.disabled) {
        backgroundColor = Color.Zircon;
        borderColor = Color.Pale;
    } else if (props.color === 'blue' && props.highlighted) {
        borderColor = Color.Blue;
    } else if (props.color === 'green' && props.highlighted) {
        borderColor = Color.Green;
    }
    if (props.checked) {
        animationState = css`
            opacity: 0;
            transform: scale(0);
        `;
    }
    return css`
        border: 1px solid ${borderColor};
        background-color: ${backgroundColor};
        ${animationState}
    `;
};

interface EmptyCircleProps {
    color: string
    checked: boolean
    highlighted: boolean
    disabled: boolean
}

export const EmptyCircle = styled.div<EmptyCircleProps>`
    z-index: 100;
    position: absolute;
    box-sizing: border-box;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    transition:
        opacity .6s cubic-bezier(0.23, 1, 0.32, 1),
        transform .6s cubic-bezier(0.23, 1, 0.32, 1),
        border-color .12s linear;
    ${getEmptyCircleStyles}
`;

const getCircleStyles = (props) => {
    let backgroundColor = Color.White;
    let circleColor = Color.Blue;
    let borderColor = Color.Pale;
    let animationState = css`
        opacity: 0;
        transform: scale(0);
    `;
    let circleSize = css`
        width: 8px;
        height: 8px;
        margin: -4px 0 0 -4px;
    `;
    if (props.disabled) {
        backgroundColor = Color.Zircon;
        circleColor = Color.Gray;
        borderColor = Color.Pale;
    } else if (props.color === 'blue' && props.highlighted) {
        borderColor = Color.Blue;
    } else if (props.color === 'green') {
        circleColor = Color.Green;
        if (props.highlighted) {
            borderColor = Color.Green;
        }
    }
    if (props.checked) {
        animationState = css`
            opacity: 1;
            transform: scale(1);
        `;
    }
    if (props.large) {
        circleSize = css`
            width: 10px;
            height: 10px;
            margin: -5px 0 0 -5px;
        `;
    }
    return css`
        border: 1px solid ${borderColor};
        background-color: ${backgroundColor};
        ${animationState};
        &::before {
            ${circleSize};
            background-color: ${circleColor};
        }
    `;
};

interface CircleProps extends EmptyCircleProps {
    large: boolean
}

export const Circle = styled.div<CircleProps>`
    z-index: 200;
    position: absolute;
    box-sizing: border-box;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    ${getCircleStyles};
    transition:
        opacity .6s cubic-bezier(0.23, 1, 0.32, 1),
        transform .6s cubic-bezier(0.23, 1, 0.32, 1),
        border-color .12s linear;
    &::before {
        position: absolute;
        content: '';
        display: block;
        top: 50%;
        left: 50%;
        border-radius: 50%;
    }
`;
