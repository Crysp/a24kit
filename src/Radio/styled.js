import styled, { css } from 'styled-components';
import { Color } from '../config';

export const Wrapper = styled.div`
    position: relative;
    display: inline-block;
    top: ${props => (props.large ? 'auto' : '2px')};
    width: ${props => (props.large ? 20 : 16)}px;
    height: ${props => (props.large ? 20 : 16)}px;
`;

export const EmptyCircle = styled.div.attrs((props) => {
    let backgroundColor = Color.White;
    let borderColor = Color.Pale;
    let animationState = css`
        opacity: 1;
        transform: scale(1);
    `;
    if (props.disabled) {
        backgroundColor = Color.Zircon;
        borderColor = Color.Pale;
    } else if (props.color === 'blue' && props.highlighted) {
        borderColor = Color.Blue;
    } else if (props.color === 'green' && props.highlighted) {
        borderColor = Color.Green;
    }
    if (props.checked) {
        animationState = css`
            opacity: 0;
            transform: scale(0);
        `;
    }
    return {
        backgroundColor,
        borderColor,
        animationState,
    };
})`
    z-index: 100;
    position: absolute;
    box-sizing: border-box;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    border: 1px solid ${props => props.borderColor};
    background-color: ${props => props.backgroundColor};
    transition:
        opacity .6s cubic-bezier(0.23, 1, 0.32, 1),
        transform .6s cubic-bezier(0.23, 1, 0.32, 1),
        border-color .12s linear;
    ${props => props.animationState}
`;

export const Circle = styled.div.attrs((props) => {
    let backgroundColor = Color.White;
    let circleColor = Color.Blue;
    let borderColor = Color.Pale;
    let animationState = css`
        opacity: 0;
        transform: scale(0);
    `;
    let circleSize = css`
        width: 8px;
        height: 8px;
        margin: -4px 0 0 -4px;
    `;
    if (props.disabled) {
        backgroundColor = Color.Zircon;
        circleColor = Color.Gray;
        borderColor = Color.Pale;
    } else if (props.color === 'blue' && props.highlighted) {
        borderColor = Color.Blue;
    } else if (props.color === 'green') {
        circleColor = Color.Green;
        if (props.highlighted) {
            borderColor = Color.Green;
        }
    }
    if (props.checked) {
        animationState = css`
            opacity: 1;
            transform: scale(1);
        `;
    }
    if (props.large) {
        circleSize = css`
            width: 10px;
            height: 10px;
            margin: -5px 0 0 -5px;
        `;
    }
    return {
        backgroundColor,
        circleColor,
        borderColor,
        animationState,
        circleSize,
    };
})`
    z-index: 200;
    position: absolute;
    box-sizing: border-box;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    border: 1px solid ${props => props.borderColor};
    background-color: ${props => props.backgroundColor};
    ${props => props.animationState};
    transition:
        opacity .6s cubic-bezier(0.23, 1, 0.32, 1),
        transform .6s cubic-bezier(0.23, 1, 0.32, 1),
        border-color .12s linear;
    &::before {
        position: absolute;
        content: '';
        display: block;
        top: 50%;
        left: 50%;
        ${props => props.circleSize};
        border-radius: 50%;
        background-color: ${props => props.circleColor};
    }
`;
