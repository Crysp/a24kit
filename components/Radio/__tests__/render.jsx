import React from 'react';
import { create } from 'react-test-renderer';
import Radio from '../index';


describe('Верстка компонента <Radio>', () => {
    it('Дефолтная', () => {
        const tree = create(<Radio />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Синего цвета', () => {
        const tree = create(<Radio color="blue" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Зеленого цвета', () => {
        const tree = create(<Radio color="green" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Увеличенный размер', () => {
        const tree = create(<Radio large />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Неактивная', () => {
        const tree = create(<Radio disabled />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Отмеченная', () => {
        const tree = create(<Radio isChecked />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Подсвеченная', () => {
        const tree = create(<Radio isHighlighted />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
