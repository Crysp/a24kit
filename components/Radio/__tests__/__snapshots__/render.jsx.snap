// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Верстка компонента <Radio> Дефолтная 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={false}
    className="sc-bwzfXH jImmSN"
    color="blue"
    disabled={false}
  />
  <div
    checked={false}
    className="sc-htpNat jBPXNz"
    color="blue"
    disabled={false}
  />
</div>
`;

exports[`Верстка компонента <Radio> Зеленого цвета 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={false}
    className="sc-bwzfXH jImmSN"
    color="green"
    disabled={false}
  />
  <div
    checked={false}
    className="sc-htpNat jNWshX"
    color="green"
    disabled={false}
  />
</div>
`;

exports[`Верстка компонента <Radio> Неактивная 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={false}
    className="sc-bwzfXH gzwCBV"
    color="blue"
    disabled={true}
  />
  <div
    checked={false}
    className="sc-htpNat bqxraA"
    color="blue"
    disabled={true}
  />
</div>
`;

exports[`Верстка компонента <Radio> Отмеченная 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={true}
    className="sc-bwzfXH cylRuW"
    color="blue"
    disabled={false}
  />
  <div
    checked={true}
    className="sc-htpNat jUKsXP"
    color="blue"
    disabled={false}
  />
</div>
`;

exports[`Верстка компонента <Radio> Подсвеченная 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={false}
    className="sc-bwzfXH jCbxnv"
    color="blue"
    disabled={false}
  />
  <div
    checked={false}
    className="sc-htpNat kdmDSI"
    color="blue"
    disabled={false}
  />
</div>
`;

exports[`Верстка компонента <Radio> Синего цвета 1`] = `
<div
  className="sc-bdVaJa cjORyY"
>
  <div
    checked={false}
    className="sc-bwzfXH jImmSN"
    color="blue"
    disabled={false}
  />
  <div
    checked={false}
    className="sc-htpNat jBPXNz"
    color="blue"
    disabled={false}
  />
</div>
`;

exports[`Верстка компонента <Radio> Увеличенный размер 1`] = `
<div
  className="sc-bdVaJa ddwLo"
>
  <div
    checked={false}
    className="sc-bwzfXH jImmSN"
    color="blue"
    disabled={false}
  />
  <div
    checked={false}
    className="sc-htpNat bgwKsE"
    color="blue"
    disabled={false}
  />
</div>
`;
